implementations/rust/ockam/ockam_api/src/authenticator/direct.rs

```diff
@@ -0,0 +1,419 @@
+use ockam_identity::{IdentityIdentifier, IdentitySecureChannelLocalInfo};
+use ockam_node::Context;
+use tracing::{trace, warn};
+use types::{AddEnroller, AddMember, Placeholder};
+
+// storage scopes:
+const ENROLLER: &str = "enroller";
+const MEMBER: &str = "member";
+
+const MAX_VALIDITY: Duration = Duration::from_secs(2 * 3600);
+
+#[derive(Debug)]
+pub struct Server<S, T> {
+    m_store: S, // member store
+    e_store: T, // enroller store
+    signer: signer::Client,
+    admin: Option<IdentityId<'static>>,
+}
+
+#[ockam_core::worker]
+impl<S, T> Worker for Server<S, T>
+where
+    S: AuthenticatedStorage,
+    T: AuthenticatedStorage,
+{
+    type Context = Context;
+    type Message = Vec<u8>;
+
+    async fn handle_message(&mut self, c: &mut Context, m: Routed<Self::Message>) -> Result<()> {
+        if let Ok(i) = IdentitySecureChannelLocalInfo::find_info(m.local_message()) {
+            let r = self.on_request(i.their_identity_id(), m.as_body()).await?;
+            c.send(m.return_route(), r).await
+        } else {
+            let mut dec = Decoder::new(m.as_body());
+            let req: Request = dec.decode()?;
+            let res = response::forbidden(&req, "secure channel required").to_vec()?;
+            c.send(m.return_route(), res).await
+        }
+    }
+}
+
+impl<S, T> Server<S, T>
+where
+    S: AuthenticatedStorage,
+    T: AuthenticatedStorage,
+{
+    pub fn new(m_store: S, e_store: T, signer: signer::Client) -> Self {
+        Server {
+            m_store,
+            e_store,
+            signer,
+            admin: None,
+        }
+    }
+
+    pub fn set_admin(&mut self, id: &IdentityIdentifier) {
+        self.admin = Some(IdentityId::new(id.key_id().to_string()))
+    }
+
+    async fn on_request(&mut self, from: &IdentityIdentifier, data: &[u8]) -> Result<Vec<u8>> {
+        let mut dec = Decoder::new(data);
+        let req: Request = dec.decode()?;
+
+        trace! {
+            target: "ockam_api::authenticator::direct::server",
+            from   = %from,
+            id     = %req.id(),
+            method = ?req.method(),
+            path   = %req.path(),
+            body   = %req.has_body(),
+            "request"
+        }
+
+        let res = match req.method() {
+            Some(Method::Post) => match req.path_segments::<2>().as_slice() {
+                // Admin wants to add an enroller.
+                ["enroller"] if self.is_admin(from) => {
+                    let add: AddEnroller = dec.decode()?;
+                    let enroller = minicbor::to_vec(Placeholder)?;
+                    self.e_store
+                        .set(ENROLLER, add.enroller().as_str().to_string(), enroller)
+                        .await?;
+                    Response::ok(req.id()).to_vec()?
+                }
+                ["enroller"] => Response::unauthorized(req.id()).to_vec()?,
+                // Enroller wants to add a member.
+                ["member"] => {
```

Comment:
Would be great if we could stick to the same conventions for paths and methods. E.g. Rest-like. This would be POST members. This is an example I think we had few people agreed on a while ago https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_api/src/identity/identity_service.rs

Reply:
How would that look here?

Reply:
This particular request would be POST members. I believe other requests have good examples in the link I posted. Also REST best practices are available on the Internet if we haven't yet covered some scenarios

Reply:
https://github.com/build-trust/ockam/commit/b436e864e43e9e8b839c5fc33ce6bdc725f5fa79